vayesta/core/types/wf/cisdtq.py

```diff
@@ -14,24 +14,181 @@ def CISDTQ_WaveFunction(mo, *args, **kwargs):
 
 class RCISDTQ_WaveFunction(wf_types.WaveFunction):
 
-    def __init__(self, mo, c0, c1, c2, c3, c4):
+    def __init__(self, mo, c0, c1, c2, c3, c4_abab, c4_abaa):
         super().__init__(mo)
         self.c0 = c0
         self.c1 = c1
         self.c2 = c2
         self.c3 = c3
-        self.c4 = c4
+        self.c4_abab = c4_abab
+        self.c4_abaa = c4_abaa
 
     def as_ccsdtq(self):
         t1 = self.c1/self.c0
         t2 = self.c2/self.c0 - einsum('ia,jb->ijab', t1, t1)
-        raise NotImplementedError
-        # TODO:
         # see also THE JOURNAL OF CHEMICAL PHYSICS 147, 154105 (2017)
-        t3 = self.c3/self.c0 # - C1*C2 - (C1^3)/3
-        t4 = self.c4/self.c0 # - C1*C3 - (C2^2)/2 - C1^2*C2 - (C1^4)/4
-        return wf_types.RCCSDTQ_WaveFunction(self.mo, t1=t1, t2=t2, t3=t3, t4=t4)
 
+        # === t3 ===
+        t3 = self.c3/self.c0
+        # As a useful intermediate, compute t2_aa from t2 (_ab), which is 
+        # just the antisymmetric permutation.
+        t2aa = t2 - t2.transpose(1,0,2,3)
+
+        t1t2 = einsum('ia, jkbc -> ijkabc', t1, t2)
+        t1t1t1 = einsum('ia,jb,kc -> ijkabc', t1, t1, t1)
+
+        t3 -= t1t2                                          
+        t3 += t1t2.transpose(0,1,2,5,4,3)
+        t3 += t1t2.transpose(2,1,0,3,4,5)                                      
```

Comment:
Question, is it possible to deduce the sign based on the permutation - and possible also the list of permutation itself?
If yes, this could prevent some bugs, which would be difficult to find otherwise

Reply:
Yes to both, and that is certainly the aim longer term when we can procedurally generate these expressions. @obackhouse  will look into this, but for the meantime, we will leave it like this and test correctness in other ways.

Reply:
for the record I do already have [a function for this in ebcc](https://github.com/BoothGroup/ebcc/blob/5b1133ce424656e79451b5565cb9381203e23c8e/ebcc/util.py#L160)

Reply:
Defo @obackhouse - we need to have a chat soon about integrating more of the ebcc functionality.

Reply:
These expressions are now algorithmically generated, so the permutations are determined algorithmically too.